driver/src/price_estimation/dexag/mod.rs

```diff
@@ -1 +1,178 @@
+{
+    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(tokens
+            .iter()
+            .filter_map(|token| {
+                let stable_coin_price = if token.symbol() == self.stable_coin.symbol {
+                    1.0
+                } else {
+                    let dexag_token = self.tokens.get(token.symbol())?;
+                    self.api.get_price(&self.stable_coin, dexag_token).ok()?
+                };
+                Some((token.id, token.get_owl_price(stable_coin_price)))
+            })
+            .collect())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::api::MockDexagApi;
+    use super::*;
+    use mockall::predicate::*;
+
+    #[test]
+    fn get_token_prices() {
+        let mut api = MockDexagApi::new();
+
+        let tokens = [
+            Token::new(1, "ETH", 18),
+            Token::new(4, "USDC", 6),
+            Token::new(5, "PAX", 18),
+            Token::new(6, "DAI", 18),
+        ];
+
+        let api_tokens = [
+            super::api::Token {
+                name: String::new(),
+                symbol: "DAI".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "ETH".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "USDC".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "PAX".to_string(),
+                address: None,
+            },
+        ];
+
+        let api_tokens_ = api_tokens.clone();
+        api.expect_get_token_list()
+            .returning(move || Ok(api_tokens_.to_vec()));
+
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[1].clone()))
+            .returning(|_, _| Ok(0.7));
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[2].clone()))
+            .returning(|_, _| Ok(1.2));
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[3].clone()))
+            .returning(|_, _| Err(anyhow!("")));
+        api.expect_get_price()
+            .with(
+                eq(api_tokens[0].clone()),
+                // Clippy claims we can remove the clone here but compilation
+                // actually fails if we follow the suggestion.
+                #[allow(clippy::redundant_clone)]
+                eq(api_tokens[0].clone()),
+            )
+            .returning(|_, _| Ok(1.0));
+
+        let client = DexagClient::with_api(api).unwrap();
+        let prices = client.get_prices(&tokens).unwrap();
+        assert_eq!(
+            prices,
+            hash_map! {
+                TokenId(1) => tokens[0].get_owl_price(0.7) as u128,
+                TokenId(4) => tokens[1].get_owl_price(1.2) as u128,
```

Comment:
nit: why is tokenId 5 missing? 

Also, would it make sense to write the exact number we are expecting rather than invoking a function? While I agree that this function is tested elsewhere, we are kind of leaking the fact that we are relying on it while we could just assert on the expected output.

Reply:
TokenId 5 is missing because the test is setup to return an error for this token. I have added a comment to explain that.

`get_owl_price` is the correct result so it makes sense to me. If for some reason this struct internally wanted to use a difference function it should still return the same result.

Reply:
For the tokenID 5 it sounds like this could be a separate test case where the test name implies that errored responses are ignored. We should only be testing one thing per test.

Reply:
Made a second test case.